source/kernel/src/services/i2c.rs

```diff
@@ -0,0 +1,685 @@
+                }
+            }
+        }
+    }
+
+    /// Obtain an `I2cClient`
+    ///
+    /// Does NOT attempt to get an [`I2cService`] handle more than once.
+    ///
+    /// Prefer [`I2cClient::from_registry`] unless you will not be spawning one
+    /// around the same time as obtaining a client.
+    pub async fn from_registry_no_retry(kernel: &'static Kernel) -> Option<Self> {
+        let handle = kernel.with_registry(|reg| reg.get::<I2cService>()).await?;
+
+        Some(I2cClient {
+            handle,
+            reply: Reusable::new_async().await,
+        })
+    }
+
+    /// Starts an I²C [`Transaction`] with the device at the provided
+    /// `addr`.
+    ///
+    /// This method begins a bus transaction with the target device. While the
+    /// returned [`Transaction`] type is held, other `I2cClients` cannot perform
+    /// bus operations; the bus is released when the [`Transaction`] is dropped.
+    ///
+    /// After starting a [`Transaction`], the [`Transaction::read`] and
+    /// [`Transaction::write`] methods are used to write to and read from the
+    /// target I²C device. See the [`Transaction`] type's
+    /// documentation for more information on how to use it.
+    pub async fn start_transaction(&mut self, addr: Addr) -> Transaction {
+        let resp = self
+            .handle
+            .request_oneshot(StartTransaction { addr, capacity: 2 }, &self.reply)
+            .await
+            .unwrap();
+        resp.body.expect("transaction should be created")
+    }
+}
+
+impl i2c::ErrorType for I2cClient {
+    type Error = I2cError;
+}
+
+impl i2c::I2c<i2c::SevenBitAddress> for I2cClient {
+    async fn transaction(
+        &mut self,
+        address: i2c::SevenBitAddress,
+        operations: &mut [i2c::Operation<'_>],
+    ) -> Result<(), Self::Error> {
+        let mut buf = {
+            // determine the maximum size operation to allocate a buffer for the
+            // transaction.
+            let len = operations
+                .iter()
+                .map(|op| match op {
+                    i2c::Operation::Read(buf) => buf.len(),
+                    i2c::Operation::Write(buf) => buf.len(),
+                })
+                .max();
+            FixedVec::new(len.unwrap_or(0)).await
+        };
+        let mut txn = self.start_transaction(Addr::SevenBit(address)).await;
+        let n_ops = operations.len();
+        for (n, op) in operations.iter_mut().enumerate() {
+            buf.clear();
+            let end = n == n_ops - 1;
+            match op {
+                i2c::Operation::Read(dest) => {
+                    let len = dest.len();
+                    let read = txn.read(buf, len, end).await?;
+                    dest.copy_from_slice(read.as_slice());
+                    buf = read;
+                }
+                i2c::Operation::Write(src) => {
+                    let len = src.len();
+                    buf.try_extend_from_slice(src)
```

Comment:
Could we check the capacity ahead of time instead of panicking?

Reply:
we allocate the buffer with enough capacity to fit the biggest `i2c::Operation`'s buffer, see https://github.com/tosc-rs/mnemos/blob/10185233cda73faea03716e609182c84aa0c422e/source/kernel/src/services/i2c.rs#L355-L366

if the `try_extend_from_slice` call fails, it means one of the following:
- `buf.clear()` decided not to clear the buffer for some reason
- the allocator decided to give us a buffer smaller than the one we asked for, just to fuck with us
- one of the `i2c::Operation`s was resized while we had an `&mut` reference to it (someone is doing something evil)

so, all of those suggest something Really Fucked Up is happening and i think it's reasonable to panic.